Npm1300 mfd driver

aasinclair opened the pull request:
Added MFD driver for nPM1300, with register read/write helper functions for use by submodules.

aasinclair committed:
drivers: mfd: npm1300: Initial version

Added an MFD driver for the nPM1300.
This driver has register access helper functions that can be used
by subsystems.  This will avoid each subsystem having to duplicate
the register access code.

Signed-off-by: Andy Sinclair <andy.sinclair@nordicsemi.no>

aasinclair committed:
drivers: regulator: npm1300: Now uses MFD register access functions

Local register read/write functions have been removed and replaced
with calls to the new MFD functions.

Signed-off-by: Andy Sinclair <andy.sinclair@nordicsemi.no>

aasinclair commented:
@gmarull Does this align with your thinking on moving register accesses to the MFD driver?

gmarull commented on drivers/regulator/Kconfig.npm1300:
```diff
@@ -6,23 +6,22 @@ config REGULATOR_NPM1300
 
-if REGULATOR_NPM1300
-
 config REGULATOR_NPM1300_COMMON_INIT_PRIORITY
 	int "nPM1300 regulator driver init priority (common part)"
 	default 75
+	depends on REGULATOR_NPM1300
 	help
 	  Init priority for the Nordic nPM1300 regulator driver (common part).
 	  It must be greater than I2C init priority.
 
 config REGULATOR_NPM1300_INIT_PRIORITY
 	int "nPM1300 regulator driver init priority"
 	default 76
+	depends on REGULATOR_NPM1300
 	help
 	  Init priority for the Nordic nPM1300 regulator driver. It must be
 	  greater than REGULATOR_NPM1300_COMMON_INIT_PRIORITY.
-
-endif
```
the guidelines here is to use `if` when the number of options is > 1, `depends on` for single options

gmarull commented on include/zephyr/drivers/mfd/npm1300.h:
```diff
@@ -0,0 +1,79 @@
+ * @{
+ */
+
+#include <errno.h>
+#include <stdint.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
```
```suggestion
#include <stddef.h>
#include <stdint.h>

#include <zephyr/device.h>
```

gmarull commented:
> @gmarull Does this align with your thinking on moving register accesses to the MFD driver?

yes, that is exactly what I had in mind.

aasinclair committed:
drivers: sensor: npm1300_charger: use MFD register functions

Local register read/write functions have been removed and replaced
with calls to the new MFD functions.

Signed-off-by: Andy Sinclair <andy.sinclair@nordicsemi.no>

aasinclair committed:
drivers: gpio: npm1300: Use MFD register functions

Local register read/write functions have been removed and replaced
with calls to the new MFD functions.

Signed-off-by: Andy Sinclair <andy.sinclair@nordicsemi.no>

gmarull commented on include/zephyr/drivers/mfd/npm1300.h:
```diff
@@ -0,0 +1,91 @@
+ * @param offset Register offset address (bits 7..0 of 16-bit address)
+ * @param data data to write
+ * @param mask mask of bits to be modified
+ * @retval Return value from i2c_write_read_dt or i2c_write_dt
```
format for retval is `@retval CODE DESC`, so maybe just:

```
@retval 0 If successful
@retval -errno In case of any bus error (see i2c_write_read_dt(), i2c_write_dt())
```

gmarull commented on include/zephyr/drivers/mfd/npm1300.h:
```diff
@@ -0,0 +1,91 @@
+
+/**
+ * @defgroup mdf_interface_npm1300 MFD NPM1300 Interface
+ * @ingroup mfd_interfaces
```
if `mfd_interfaces` group defined? If not, add it to `doc/_doxygen/groups.dox` since there's no "mfd.h" top level header

gmarull reviewed (commented):
lgtm, a couple of Doxygen nitpicks

henrikbrixandersen commented on drivers/mfd/mfd_npm1300.c:
```diff
@@ -0,0 +1,85 @@
+	return i2c_write_dt(&config->i2c, buff, sizeof(buff));
+}
+
+int mfd_npm1300_reg_update(const struct device *dev, uint8_t base, uint8_t offset, uint8_t data,
+			   uint8_t mask)
+{
+	uint8_t reg;
+	int ret;
+
+	ret = mfd_npm1300_reg_read(dev, base, offset, &reg);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	reg = (reg & ~mask) | (data & mask);
+
+	return mfd_npm1300_reg_write(dev, base, offset, reg);
+}
```
Doesn't this read-modify-write require locking/exclusivity? Or does the architecture of the "sub"-drivers ensure, that no driver (thread) will modify the same register between the read/write calls?

gmarull replied:
I guess it's similar to `i2c_reg_update_byte`, where no bus locking is in place. So all these drivers are for now non-thread safe. The MFD layer could introduce optional bus locking, though.

henrikbrixandersen replied:
Right, but for `i2c_reg_update_byte()` there is usually just one consumer per I2C device. Here we have multiple consumers.

aasinclair replied:
I think multi-threading is going to be quite likely going forward, so I'll add locking to the MFD driver.

henrikbrixandersen reviewed (changes requested):
Is the mutex lock/unlock really needed for the simple read/write cases? Wouldn't the underlying I2C controller ensure exclusivity there?

aasinclair commented:
ping @gmarull
